[11.x] Add fluent RegExp validation class rule

shaedrich opened the pull request:
Similar to #54488, #53465, #54425, #54517, and #54519

# Examples
## Simplest case
```php
use Illuminate\Validation\Rule;

public function rules()
{
    return [
        'foo' => [
            'required',
            Rule::regex('/[a-z]/'), // results in 'regex:/[a-z]/'
        ],
    ];
}
```

## Negation
```php
use Illuminate\Validation\Rule;

public function rules()
{
    return [
        'foo' => [
            'required',
            Rule::regex('/[a-z]/')->not(), // results in 'not_regex:/[a-z]/'
        ],
    ];
}
```

## Flags
```php
use Illuminate\Validation\Rule;

public function rules()
{
    return [
        'foo' => [
            'required',
            Rule::regex('/[a-z]/i'), // results in 'regex:/[a-z]/i'
            // or Rule::regex('/[a-z]/', ['i'])
            // or Rule::regex('/[a-z]/')->flags(['i'])
        ],
    ];
}
```

### Modify flags
```php
use Illuminate\Validation\Rule;

public function rules()
{
    $rule = Rule::regex('/[a-z]/i');
    $rule->flag('g');
    return [
        'foo' => [
            'required',
            $rule, // results in 'regex:/[a-z]/gi'
        ],
    ];
}
```

shaedrich committed:
Add fluent RegExp validation class rule

shaedrich committed:
Fix import paths

shaedrich committed:
Make argument optional

shaedrich committed:
Fix typos

shaedrich committed:
Fix class property name

shaedrich committed:
Use str_split() instead of explode()

shaedrich committed:
Spread array

shaedrich committed:
Fix expectation

AndrewMast commented:
I like this suggestion! I have a few comments that I feel like will help it be inline with the usual Laravel code that I will add in a review/comments.

AndrewMast commented on src/Illuminate/Validation/Rules/RegExp.php:
```diff
@@ -0,0 +1,88 @@
+
+class RegExp implements Stringable
+{
+    protected string $regExp;
```
I feel like this should be named `$expression`, but not really important.

shaedrich commented:
@AndrewMast Sure, feedback is always welcome 👍🏻

AndrewMast commented on src/Illuminate/Validation/Rules/RegExp.php:
```diff
@@ -0,0 +1,88 @@
+        $this->flags(...$flags);
+    }
+
+    public function active(bool $active)
```
What is the purpose of this method? Do we even have an `$active` prop? I may just be forgetting what is in `Stringable` though.

shaedrich replied:
Good catch! I accidentally copied it from https://github.com/laravel/framework/pull/54519/files#diff-4869bb9b60c679381b374e53caf029cc40da95aaadc62efb7e1e15795489bd90R28-R33

AndrewMast commented on src/Illuminate/Validation/Rules/RegExp.php:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * @param  string[]|\Illuminate\Contracts\Support\Arrayable|null  $flags
+     */
+    public function flags($flags = null)
```
I feel like accepting `null` here and in the constructor is counterintuitive, especially if it is an invalid input. What if someone calls `->flags(null)`, wouldn't your code assign `$this->flags` to be `[null]`?

I propose `array|Arrayable $flags = []` for this method and the constructor. Or, to allow `func_get_args()` to work, `$flags = []`.

shaedrich replied:
The class essentially has two methods:
* `setFlags()`: Overwrites/replaces the existing flags with a set of new ones (here, it's called `flags()` due to naming conventions)
* `addFlag()`: Adds a flag to the set of existing flags without replacing them (here, it's called `flag()` due to naming conventions)

AndrewMast commented on src/Illuminate/Validation/Rules/RegExp.php:
```diff
@@ -0,0 +1,88 @@
+                ':',
+                $this->regExp,
+            )
+            ->when($this->flags, function (SupportStringable $str) {
```
What if we get all of the flags from the expression when we create the instance? This way if we call `->flags('i', 'g', ...)` (which current replaces *all* flags with the flags provided), it will not keep any of the flags originally in the expression. Then we can just add `implode('', $this->flags)` to the `->append(...)` here.

AndrewMast replied:
Maybe we could do something like this for the constructor:

```php
public function __construct(string $expression, array|Arrayable $extraFlags = [])
{
    $currentFlags = str_split(Str::of($expression)->afterLast('/'));

    if ($extraFlags instanceof Arrayable) {
        $extraFlags = $extraFlags->toArray();
    }
    
    $this->expression = Str::of($expression)->beforeLast('/')->append('/')->toString();
    $this->flags = array_unique([...$currentFlags, ...$extraFlags]);
}
```

AndrewMast replied:
That's what I assumed! Now with the updated constructor code, it makes much more sense to me. Good work!

shaedrich replied:
> Good work!

Primarily due to your feedback 👍🏻 Thanks a lot!

shaedrich committed:
Remove unused method

shaedrich committed:
fix(validation): :pencil2: Fix typo

shaedrich committed:
refactor(validation): :recycle: Move flags logic to constructor

Based on https://github.com/laravel/framework/pull/54521/files/daffcc29b48dcc7aa083e10e803a4242eb5ed5de#r1947924894

Co-authored-by: Andrew Mast <andrew@andrewmast.com>

shaedrich committed:
refactor(validation): :truck: Rename $regExp to $expression

see https://github.com/laravel/framework/pull/54521/files/daffcc29b48dcc7aa083e10e803a4242eb5ed5de#r1947921675

Co-authored-by: Andrew Mast andrew@andrewmast.com

shaedrich committed:
refactor(validation): :coffin: Remove unused import

shaedrich committed:
style(validation): :art: Fix indent

AndrewMast commented on src/Illuminate/Validation/Rules/RegExp.php:
```diff
@@ -0,0 +1,82 @@
+     */
+    public function __construct(string $expression, array|Arrayable $extraFlags = [])
+    {
+        $this->expression = $expression;
```
I think we can take this line out? We assign the expression again on line 36

shaedrich committed:
refactor(validation): :coffin: Remove redunant line

see https://github.com/laravel/framework/pull/54521#discussion_r1947943119

Co-authored-by: Andrew Mast <andrew@andrewmast.com>

shaedrich committed:
refactor(validation): :label: Improve typing of flags() method

see https://github.com/laravel/framework/pull/54521#discussion_r1947922366

Co-authored-by: Andrew Mast <andrew@andrewmast.com>

shaedrich committed:
refactor(validation): :white_check_mark: Adjust test

Follow-up to 4e18d64

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If applicable, please consider releasing your code as a package so that the community can still take advantage of your contributions!
